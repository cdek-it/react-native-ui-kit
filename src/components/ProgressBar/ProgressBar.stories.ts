import type { Meta, StoryObj } from '@storybook/react'

import { ProgressBar } from './ProgressBar'

const meta: Meta<typeof ProgressBar> = {
  title: 'ProgressBar',
  component: ProgressBar,
  args: { showValue: false, value: 20, style: { height: 7 } },
  parameters: {
    notes: `
      Компонент ProgressBar

      <ProgressBar
        value={20}
        showValue={false}
        style={{ height: 10, width: 200 }}
      />
    `,
  },
}

export default meta

const ProgressBarStory: StoryObj<typeof ProgressBar> = {}

export { ProgressBarStory as ProgressBar }
