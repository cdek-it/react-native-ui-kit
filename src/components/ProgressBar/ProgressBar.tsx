import { memo, useCallback, useEffect, useMemo, useState } from 'react'
import {
  View,
  Text,
  type LayoutChangeEvent,
  type StyleProp,
  type ViewProps,
  type ViewStyle,
} from 'react-native'
import Animated, { useSharedValue, withTiming } from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

/**
 * Тип свойств компонента ProgressBar
 * @see ProgressBar
 */
export interface ProgressBarProps extends Pick<ViewProps, 'testID'> {
  /**
   * Значение прогресса от 0 до 100
   */
  value: number
  /**
   * Отображение значения прогресса
   */
  showValue?: boolean
  /**
   * Стилизация компонента ProgressBar
   */
  style?: StyleProp<ProgressBarStyle>
}

/**
 * Компонент ProgressBar
 * @param value - значение прогресса от 0 до 100
 * @param showValue - отображение значения прогресса
 * @param style - стилизация компонента ProgressBar
 * @see ProgressBarProps
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-5024&m=dev
 */
export const ProgressBar = memo<ProgressBarProps>(
  ({ value: propsValue, showValue = false, style }) => {
    const styles = useStyles()

    const [containerWidth, setContainerWidth] = useState(0)
    const indicatorWidth = useSharedValue(0)

    const onLayout = useCallback(
      ({
        nativeEvent: {
          layout: { width },
        },
      }: LayoutChangeEvent) => {
        setContainerWidth(width)
      },
      []
    )

    const value = useMemo(() => {
      if (propsValue > 100) {
        return 100
      }

      if (propsValue < 0) {
        return 0
      }

      return propsValue
    }, [propsValue])

    useEffect(() => {
      indicatorWidth.value = withTiming(containerWidth * (value / 100))
    }, [containerWidth, indicatorWidth, value])

    return (
      <View
        style={[
          styles.container,
          style,
          showValue && styles.containerShowValue,
        ]}
        onLayout={onLayout}
      >
        <Animated.View style={[styles.indicator, { width: indicatorWidth }]}>
          {showValue ? (
            <Text style={styles.indicatorText}>{value} %</Text>
          ) : null}
        </Animated.View>
      </View>
    )
  }
)

const useStyles = makeStyles(({ theme, typography }) => ({
  container: {
    borderRadius: 100,
    backgroundColor: theme.Misc.ProgressBar.progressBarBg,
    overflow: 'hidden',
  },

  indicator: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: theme.Misc.ProgressBar.progressBarValueBg,
    height: '100%',
    overflow: 'hidden',
  },

  containerShowValue: { height: theme.Misc.ProgressBar.progressBarHeight },

  indicatorText: {
    fontSize: typography.Size['text-xs'],
    textAlign: 'center',
    color: theme.Misc.ProgressBar.progressBarValueTextColor,
  },
}))

/**
 * Стиль компонента ProgressBar
 * @see ProgressBar
 */
export type ProgressBarStyle = Pick<
  ViewStyle,
  | 'height'
  | 'width'
  | 'maxWidth'
  | 'minWidth'
  | 'maxHeight'
  | 'minHeight'
  | 'flex'
  | 'alignSelf'
  | 'position'
  | 'top'
  | 'right'
  | 'left'
  | 'bottom'
  | 'margin'
  | 'marginBottom'
  | 'marginEnd'
  | 'marginHorizontal'
  | 'marginLeft'
  | 'marginRight'
  | 'marginStart'
  | 'marginTop'
  | 'marginVertical'
>
