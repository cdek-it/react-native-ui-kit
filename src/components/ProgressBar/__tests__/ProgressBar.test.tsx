import { render } from '@testing-library/react-native'

import { ProgressBar, type ProgressBarProps } from '../ProgressBar'

describe('ProgressBar component test', () => {
  const progressBarSnapshoteCases = generatePropsCombinations<ProgressBarProps>(
    {
      value: [-50, 0, 50, 100, 150],
      showValue: [true, false],
      style: [
        { height: 10, width: 100 },
        { height: 20, width: 150 },
      ],
    }
  )

  test.each(progressBarSnapshoteCases)(
    'value - $value, showValue - $showValue, style - $style',
    (props) => {
      const renderedProgressBar = render(<ProgressBar {...props} />)

      expect(renderedProgressBar.toJSON()).toMatchSnapshot()
    }
  )
})
