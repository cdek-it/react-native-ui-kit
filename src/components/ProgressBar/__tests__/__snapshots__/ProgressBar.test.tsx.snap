// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component test value - -50, showValue - false, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - -50, showValue - false, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - -50, showValue - true, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      0
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - -50, showValue - true, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      0
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 0, showValue - false, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 0, showValue - false, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 0, showValue - true, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      0
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 0, showValue - true, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      0
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 50, showValue - false, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 50, showValue - false, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 50, showValue - true, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      50
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 50, showValue - true, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      50
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 100, showValue - false, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 100, showValue - false, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 100, showValue - true, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      100
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 100, showValue - true, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      100
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 150, showValue - false, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 150, showValue - false, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      false,
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  />
</View>
`;

exports[`ProgressBar component test value - 150, showValue - true, style - {"height": 10, "width": 100} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 10,
        "width": 100,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      100
       %
    </Text>
  </View>
</View>
`;

exports[`ProgressBar component test value - 150, showValue - true, style - {"height": 20, "width": 150} 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 100,
        "overflow": "hidden",
      },
      {
        "height": 20,
        "width": 150,
      },
      {
        "height": 14,
      },
    ]
  }
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#188700",
        "height": "100%",
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 0,
      }
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "textAlign": "center",
        }
      }
    >
      100
       %
    </Text>
  </View>
</View>
`;
